metadata/src/test/java/org/apache/kafka/controller/QuorumControllerMetricsTest.java

```diff
@@ -0,0 +1,82 @@
+ * limitations under the License.
+ */
+
+package org.apache.kafka.controller;
+
+import com.yammer.metrics.core.MetricsRegistry;
+import org.junit.jupiter.api.Test;
+
+import java.util.Arrays;
+import java.util.Collections;
+import java.util.HashSet;
+import java.util.Set;
+import java.util.stream.Collectors;
+
+import static org.junit.jupiter.api.Assertions.assertEquals;
+
+public class QuorumControllerMetricsTest {
+    @Test
+    public void testKafkaControllerMetricNames() {
+        String expectedGroup = "kafka.controller";
+        String expectedType = "KafkaController";
+        Set<String> expectedMetricNames = new HashSet<>(Arrays.asList(
+            "ActiveControllerCount",
+            "GlobalTopicCount",
+            "GlobalPartitionCount",
+            "OfflinePartitionCount",
+            "PreferredReplicaImbalanceCount"));
+        Set<String> missingMetrics = getMissingMetricNames(expectedMetricNames, expectedGroup, expectedType);
+        assertEquals(Collections.emptySet(), missingMetrics, "Expected metrics did not exist");
+    }
+
+    @Test
+    public void testControllerEventManagerMetricNames() {
+        String expectedGroup = "kafka.controller";
+        String expectedType = "ControllerEventManager";
+        Set<String> expectedMetricNames = new HashSet<>(Arrays.asList(
+            "EventQueueTimeMs",
+            "EventQueueProcessingTimeMs"));
+        Set<String> missingMetrics = getMissingMetricNames(expectedMetricNames, expectedGroup, expectedType);
+        assertEquals(Collections.emptySet(), missingMetrics, "Expected metrics did not exist");
+    }
+
+    private static Set<String> getMissingMetricNames(
+        Set<String> expectedMetricNames, String expectedGroup, String expectedType) {
+        MetricsRegistry registry = new MetricsRegistry();
+        new QuorumControllerMetrics(registry); // populates the registry
+        Set<String> foundMetricNames = expectedMetricNames.stream().filter(expectedMetricName ->
+            registry.allMetrics().keySet().stream().anyMatch(metricName -> {
+                if (metricName.getGroup().equals(expectedGroup) && metricName.getType().equals(expectedType)
+                    && metricName.getScope() == null && metricName.getName().equals(expectedMetricName)) {
+                    // It has to exist AND the MBean name has to be correct;
+                    // fail right here if the MBean name doesn't match
+                    String expectedMBeanPrefix = expectedGroup + ":type=" + expectedType + ",name=";
+                    assertEquals(expectedMBeanPrefix + expectedMetricName, metricName.getMBeanName());
+                    return true; // the expected metric name exists and the associated MBean name matches
+                } else {
+                    return false; // this one didn't match
+                }
+            })).collect(Collectors.toSet());
+        if (foundMetricNames.size() == expectedMetricNames.size()) {
```

Comment:
nit: is the size check sufficient? Any harm skipping this check and following the other branch in all cases?

Reply:
The size check is sufficient, but there's really no significant harm in following the other branch in all cases (just unnecessarily copying a small set), so I'll go with the cleaner single solution as you suggest.